enhancements/installer/custom-dns.md

```diff
@@ -0,0 +1,574 @@
+already in use for on-prem platforms but differs from them in its use of the cloud
+load balancer.
+
+### Workflow Description
+
+1. For an OpenShift administrator to use a custom DNS solution rather than
+letting OpenShift automatically configure the cloud provided, DNS solution,
+they would have to do 2 things:
+  - Pre-configure entries in their custom DNS soltion with entries for the
+API and Ingress services
+  - Let the OpenShift Installer know that they would like to use a custom
+DNS solution and skip configuration of the cloud provided DNS solution.
+
+2. The Installer next needs to validate the presence of this custom DNS and
+its state by using an utility like `dig` to determine if the API service is
+reachable and its IPs are within each of the public subnets provided as part
+of installer configuration (also called as install-config). [TODO: can we use
+teh same validation method for the Ingress service configuration?]
+
+3. When the custom DNS configuration mentioned in #1 is present, skip generating
+terraform configuration for DNS configuration in Route53.
+
+4. The Ingress Controller configures and manages the DNS configuration for the
+Ingress service. The Ingress Controller needs to be made aware of the presence
+of the custom DNS via the DNS manifest so that it skips configuring the cloud
+DNS for the Ingress service.
+
+5. The Ingress Controller continues managing the Load Balancer configuration for
+the Ingress service.
+
+6. When the custom DNS solution is enabled, installer tries to resolve the
+API Server URL to obtain the IP adderesses within each subnet in the machine
+network and validate them.
+
+7. The API-Int Server would be configured to resolve to the IP addresses provided
+by the user. (It could be configured to resolve to the same IP addresses
+as the API Server.) It is the responsibilty of the Installer to provide these
+IP addresses to the machine-config-operator (MCO) via the Infrastructure CR.
+
+8. The Installer adds the above API-Int Server IPs to the AWSPlatformStatus
+struct within the Infrastructure Status. A new field needs to be added to
+AWSPlatformStatus for this.
+
+9. During the bootstrap phase, it appears that only the API and Ingress services
+need to be resolvable. So, no additional changes are needed during the bootstrap
+phase to support this feature.
+
+10. The MCO is also responsible for generating manifests for running CoreDNS on
+the master and worker nodes using information provided in the Infrastructure CR.
+
+### API Extensions
+
+#### api.dns [https://github.com/openshift/api/blob/master/config/v1/types_dns.go]
+
+We need a way to inform the Ingress Controller to not configure the DNS records
+for the Ingress service in the cloud DNS when the custom DNS has already been
+configured. So, adding a string config to the DNSSpec of the DNS CR to
+indicate whether the DNS configuration has already been handled externally
+should take care of that.
```

Comment:
There is a solution for this in UPI: delete the entries in the DNS cluster object.

Reply:
I did look into that while doing my research.

In the regular IPI case, the Ingress Controller would/(should ?) consider the absence of DNS zones as an error/mis-configuration. So in the custom DNS case, I wanted to make it explicit that DNS zones are missing because custom DNS is enabled.

@Miciah and @patrickdillon I can fall back to just manipulating the DNS zones for custom DNS if Ingress Controller doesn't consider the absence of DNS zones as an error in an IPI installation. (I understand why this won't be an error in an UPI install.)

Reply:
> In the regular IPI case, the Ingress Controller would/(should ?) consider the absence of DNS zones as an error/mis-configuration. So in the custom DNS case, I wanted to make it explicit that DNS zones are missing because custom DNS is enabled.

> @Miciah and @patrickdillon I can fall back to just manipulating the DNS zones for custom DNS if Ingress Controller doesn't consider the absence of DNS zones as an error in an IPI installation. (I understand why this won't be an error in an UPI install.)

AFAIK the ingress controller would be fine if we omitted the DNS zone entries, that is a valid configuration; but I think the issue is that other cluster operators depend on having ingress, so we need a solution either inside or outside the cluster to provide that ingress during or pre-install. Right? Ingress is required for install and can't be managed day 2?

Reply:
> Right? Ingress is required for install and can't be managed day 2?

Miciah's [comment](https://github.com/openshift/enhancements/pull/1276#discussion_r1018542353) below confirmed my question. Auth and console depend on ingress.